mpqp/tools/theoretical_simulation.py

```diff
@@ -0,0 +1,210 @@
+from __future__ import annotations
+
+import numpy as np
+import numpy.typing as npt
+from matplotlib import pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D
+from scipy.spatial.distance import jensenshannon
+
+from mpqp import QCircuit
+from mpqp.core.instruction.measurement.basis_measure import BasisMeasure
+from mpqp.execution import AWSDevice
+from mpqp.execution.devices import AvailableDevice
+from mpqp.execution.runner import _run_single  # pyright: ignore[reportPrivateUsage]
+from mpqp.noise import Depolarizing
+from mpqp.noise.noise_model import NoiseModel
+
+
+def theoretical_probs(
+    circ: QCircuit,
+) -> npt.NDArray[np.float32]:
+    """Computes the theoretical probabilities of a (potentially noisy circuit
+    execution.
+
+    Args:
+        circ: The circuit to run.
+
+    Returns:
+        The probabilities corresponding to each basis state.
+    """
+    d: int = 2 ** (circ.nb_qubits)
+
+    state = np.zeros((d, d), dtype=np.complex64)
+    state[0, 0] = 1
+
+    for gate in circ.get_gates():
+        g = gate.to_matrix(circ.nb_qubits).astype(np.complex64)
+        state = g @ state @ g.T.conj()
+        for noise in circ.noises:
+            if (
+                len(noise.gates) == 0
+                or type(gate) in noise.gates
+                and gate.connections().issubset(noise.targets)
+            ):
+                state = sum(
+                    (
+                        k @ state @ k.T.conj()
+                        for k in noise.to_adjusted_kraus_operators(
+                            gate.connections(), circ.nb_qubits
+                        )
+                    ),
+                    start=np.zeros((d, d), dtype=np.complex64),
+                )
+
+    connected_qubits = set().union(*[gate.connections() for gate in circ.get_gates()])
```

Comment:
Since we call two times `circ.get_gates()` here and in line 35, it can be interesting to call it only once, and store it ?

Reply:
good point